engine/runtime/src/main/scala/org/enso/compiler/context/SuggestionDiff.scala

```diff
@@ -0,0 +1,130 @@
+package org.enso.compiler.context
+import org.enso.polyglot.Suggestion
+import org.enso.polyglot.data.{These, Tree}
+import org.enso.polyglot.runtime.Runtime.Api
+
+object SuggestionDiff {
+
+  /** Compute difference between the two suggestion trees.
+    *
+    * @param prev the tree before
+    * @param current the tree after
+    * @return the tree with updates
+    */
+  def compute(
+    prev: Tree.Root[Suggestion],
+    current: Tree.Root[Suggestion]
+  ): Tree.Root[Api.SuggestionUpdate] =
+    Tree
+      .zipBy(prev, current)(compare)
+      .map(diff)
+
+  /** Compare two suggestions for equality. */
+  private def compare(a: Suggestion, b: Suggestion): Boolean =
+    a.name == b.name &&
+    a.module == b.module &&
+    Suggestion.Kind(a) == Suggestion.Kind(b) &&
+    Suggestion.SelfType(a) == Suggestion.SelfType(b)
+
+  /** Build the suggestion update from the result of tree joining operation.
+    *
+    * @param elem the result of joining two trees
+    * @return the suggestion update
+    */
+  private def diff(elem: These[Suggestion, Suggestion]): Api.SuggestionUpdate =
+    elem match {
+      case These.Here(e) =>
+        Api.SuggestionUpdate(e, Api.SuggestionAction.Remove())
+      case These.There(e) =>
+        Api.SuggestionUpdate(e, Api.SuggestionAction.Add())
+      case These.Both(e1: Suggestion.Atom, e2: Suggestion.Atom) =>
+        diffAtoms(e1, e2)
+      case These.Both(e1: Suggestion.Method, e2: Suggestion.Method) =>
+        diffMethods(e1, e2)
+      case These.Both(e1: Suggestion.Function, e2: Suggestion.Function) =>
+        diffFunctions(e1, e2)
+      case These.Both(e1: Suggestion.Local, e2: Suggestion.Local) =>
```

Comment:
But they are of the same type, `These.Both[Suggestion, Suggestion]`. I don't see how to make it more typesafe.

Reply:
You can make it more typesafe by defining `These.Both` to take a single type parameter.

Reply:
It kinda limits what we can do with `These` (like we can only zip collection of the same type), and I wanted to avoid that. 
But it still suits our purposes. 
So, ok then, I'll make this change

Reply:
Nah, `These.Both[A](a: A, b: A)` didn't work. Since `Suggestion` is an ADT, it can be created using different constructors, and this pattern match is still not exhaustive.

Reply:
Ah okay.